_posts/2020-05-07-removing-capybara-webkit.md

```diff
@@ -0,0 +1,135 @@
+
+```diff
+  def select_from_chosen(item_text, options)
+-    field = find('#' + options[:from], visible: false)
+-    option_value = page.evaluate_script("$(\"##{field[:id]} option:contains('#{item_text}')\").val()")
+-    page.execute_script("value = ['#{option_value}']\; if ($('##{field[:id]}').val()) {$.merge(value, $('##{field[:id]}').val())}")
+-    option_value = page.evaluate_script("value")
+-    page.execute_script("$('##{field[:id]}').val(#{option_value})")
+-    page.execute_script("$('##{field[:id]}').trigger('liszt:updated').trigger('change')")
+
++    field = find_field(options[:from], visible: false)
++    find("##{field[:id]}_chzn").click
++    find("##{field[:id]}_chzn ul.chzn-results li", text: item_text).click
+  end
+```
+
+So we not only updated the gem, we also improved the test suite by not doing obscure Javascript calls.
+
+# Updating Docker Container
+
+Now that our tests are passing we can update the `Dockerfile` removing the commands that installed QT's libraries `qt5-default` and `libqt5webkit5-dev` (that will depend on the OS your are using, we have a Debian image so we just removed those packes from the `apt-get` command).
+
+Then we make sure we have Mozilla Firefox or Google Chrome installed. It was easier to install Firefox since the `firefox-esr` package was already available on the Debian Buster image we were using. You can use a Docker image which includes other browsers or install Google Chrome if you want to.
+
+So we changed the default driver to use Firefox:
+
+```ruby
+Capybara.javascript_driver = :headless_firefox
+```
+
+Now our tests can be run inside Docker and we are all green.
+
+# Updating CircleCI Config
+
+The last part of the process was to make sure the tests pass on the CI system. We only needed to make sure Firefox was available when running the test, so the easiest solution for this was to use a CircleCI Ruby image and use the correct [variant](https://circleci.com/docs/2.0/circleci-images/#language-image-variants) so it also includes most used browsers:
+
+```diff
+- - image: circleci/ruby:2.4.10-buster-node
++ - image: circleci/ruby:2.4.10-buster-node-browsers
+```
+
+> We did have some problems related to caching, so you may want to try to retry your CI job without caching just to make sure if you see something failing
+
+## Conclusion
+
+We finally got an old dependency removed (QT) replacing `capybara-webkit` with a more modern solution.
+
+It makes it easier for new developers to join the project now that it has less extra requirements. It will also help when updating the Rails version, since newer Rails versions adds `webdrivers` gem when creating new apps by default.
```

Comment:
```suggestion
It makes it easier to maintain our Rails application and it gets us closer to "the Rails way". It will help us when upgrading Rails because more recent versions of Rails rely on the `webdrivers` gem by default.
```

Reply:
I don't think it makes it easier to onboard new developers, it's pretty much the same. It replaces capybara-webkit+qt with webdrivers+firefox. So _technically_ there are no fewer requirements...

Reply:
Maybe I didn't express that right. I mean that QT is a complicated dependency (https://github.com/thoughtbot/capybara-webkit/wiki/Installing-Qt-and-compiling-capybara-webkit) that you'll have to install just to run the tests suite, and firefox (or chrome) is usually there for any web developer at least for manual testing purposes, I'm not sure if that counts or not as a requirement.

Technically it may be the same amount of requirements, but I thought the whole point was to remove that qt dependency because it was harder to setup so I do think it helps to onboard new developers, maybe not because of fewer requirements but because of `easier` requirements?

What do you think?

Reply:
@arielj Got it! Then maybe you can add a note about that: 

> Installing QT is quite complicated (https://github.com/thoughtbot/capybara-webkit/wiki/Installing-Qt-and-compiling-capybara-webkit). Firefox (or Chrome) are usually already installed in any web developer's environment.

Reply:
Thanks, I only changed the `Rails rely on the...` phrase, cause it gives the impression that you need that gem for newer Rails version but it's just the rails way and can be changed if someone wants to.